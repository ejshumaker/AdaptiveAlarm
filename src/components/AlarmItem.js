/*
 * Used to display all alarms withing AlarmListScreen.js
 * @weinoh 04-10-2019
 */
/* eslint-disable camelcase */
import React, { Component } from 'react';
import {
  View, Text, TouchableOpacity, Switch, StyleSheet,
} from 'react-native';
import PropTypes from 'prop-types';
import { GlobalStyles, Colors } from '../constants';

const styles = StyleSheet.create({
  alarmRow: {
    // flex: 8,
    flexDirection: 'row',
    justifyContent: 'space-between',
    borderBottomWidth: 0.5,
    borderBottomColor: Colors.white,
    marginVertical: 0,
    width: '100%',
  },
  alarmInfoColumn: {
    // flex: 0.5,
    flexDirection: 'column',
    justifyContent: 'flex-end',
    marginVertical: 8,
  },
  switchColumn: {
    margin: 0,
    marginRight: 13,
    justifyContent: 'space-around',
  },
});

class AlarmItem extends Component {
  // on press we should call the navigator to go to the edit alarm page.
  handlePress = async () => {
    const {
      alarmId, navigate,
    } = this.props;
    // navigate to edit alarm page with params TODO: SWITCH TO EDIT ALARM SCREEN
    navigate('CreateAlarm', {
      alarmId,
    });
  };

  // on toggle of a specific alarm we should call the alarm set status prop function
  // TODO: HANDLE THE ALARM FIRING RIGHT AFTER WE TURN IT ON IF IT IS IN THE PAST
  handleAlarmToggle = async () => {
    const {
      alarmId, toggleAlarm, alarm,
    } = this.props;
    // alarm.isActive = !alarm.isActive;
    // call the toggle alarm function from alarmlistscreen (fire off action)
    // TODO: get this to update correctly.
    const toggleVal = !alarm.isActive;
    await toggleAlarm(alarmId, toggleVal);
    this.render();
  }

  displayDays() {
    const { alarm } = this.props;
    let { days } = alarm;
    days = days || {};
    let dayString = '';

    if (days.sun) dayString += ' S';
    if (days.mon) dayString += ' M';
    if (days.tue) dayString += ' T';
    if (days.wed) dayString += ' W';
    if (days.thu) dayString += ' Th';
    if (days.fri) dayString += ' F';
    if (days.sat) dayString += ' Sa';

    if (dayString === '') dayString = ' None';

    return dayString;
  }

  render() {
    const { alarm } = this.props;
    if (alarm === undefined) return null;
    return (
      <TouchableOpacity onPress={this.handlePress}>
        <View
          style={styles.alarmRow}
        >
          <View style={styles.alarmInfoColumn}>
            <Text
              style={[GlobalStyles.h2,
                { color: alarm.isActive ? Colors.white : Colors.darkGray }]}
            >
              {alarm.arrivalTime}
            </Text>
            <View style={{ marginTop: 5, flexDirection: 'row' }}>
              <Text style={[GlobalStyles.paragraph, { justifyContent: 'flex-start', color: alarm.isActive ? Colors.white : Colors.darkGray }]}>
              Repeat:
              </Text>
              <Text style={[GlobalStyles.paragraph, { justifyContent: 'flex-end', color: alarm.isActive ? Colors.primary : Colors.darkGray, fontWeight: 'bold' }]}>
                {this.displayDays()}
              </Text>
            </View>
          </View>
          <View style={styles.switchColumn}>
            <Switch
              onValueChange={this.handleAlarmToggle}
              value={alarm.isActive}
              style={styles.switchRow}
            />
          </View>

        </View>
      </TouchableOpacity>
    );
  }
}

export default AlarmItem;

AlarmItem.propTypes = {
  toggleAlarm: PropTypes.func.isRequired,
  alarm: PropTypes.object.isRequired, // eslint-disable-line
  alarmId: PropTypes.string.isRequired,
  navigate: PropTypes.func.isRequired,
};
