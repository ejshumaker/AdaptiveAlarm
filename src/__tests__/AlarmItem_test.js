// app/src/components/__tests__/Login-test.js
/* eslint-disable no-undef */
/* eslint-disable import/first */
import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import AlarmItem from '../components/AlarmItem';

Enzyme.configure({ adapter: new Adapter() });
describe('Alarm Item', () => {
  let wrapper;
  // our mock login function to replace the one provided by mapDispatchToProps
  const toggleAlarmMock = jest.fn();
  const alarmId = '1';
  const navigate = jest.fn();
  let days = {
    sun: true,
    mon: false,
    tue: true,
    wed: true,
    thu: false,
    fri: true,
    sat: false,
  };
  let alarm = {
    arrivalTime: '8:00 AM',
    isActive: true,
    days,
  };

  beforeEach(() => {
    // pass the mock function as the login props
    days = {
      sun: true,
      mon: false,
      tue: true,
      wed: true,
      thu: false,
      fri: true,
      sat: false,
    };
    alarm = {
      arrivalTime: '8:00 AM',
      isActive: true,
      days,
    };
    jest.clearAllMocks();
  });

  it('test alarm item matches snapshot alarm is active', () => {
    alarm.isActive = true;
    wrapper = shallow(<AlarmItem
      toggleAlarm={toggleAlarmMock}
      alarmId={alarmId}
      navigate={navigate}
      alarm={alarm}
    />);
    expect(wrapper).toMatchSnapshot();
  });

  it('test alarm item matches snapshot alarm is inactive', () => {
    alarm.isActive = false;
    wrapper = shallow(<AlarmItem
      toggleAlarm={toggleAlarmMock}
      alarmId={alarmId}
      navigate={navigate}
      alarm={alarm}
    />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call the handle press function and navigate to CreateAlarm', () => {
    alarm.isActive = true;
    wrapper = shallow(<AlarmItem
      toggleAlarm={toggleAlarmMock}
      alarmId={alarmId}
      navigate={navigate}
      alarm={alarm}
    />);
    wrapper.find('TouchableOpacity').simulate(
      'press',
      { preventDefault() {} },
    );
    expect(navigate.mock.calls.length).toBe(1);
    expect(navigate.mock.calls[0][0]).toEqual('CreateAlarm');
  });

  it('should call the sign in action', () => {
    alarm.isActive = true;
    wrapper = shallow(<AlarmItem
      toggleAlarm={toggleAlarmMock}
      alarmId={alarmId}
      navigate={navigate}
      alarm={alarm}
    />);
    wrapper.find('TouchableOpacity').simulate(
      'press',
      { preventDefault() {} },
    );
    expect(navigate.mock.calls.length).toBe(1);
  });

  it('should contain None as no days are active', () => {
    alarm.days = {
      sun: false,
      mon: false,
      tue: false,
      wed: false,
      thu: false,
      fri: false,
      sat: false,
    };
    alarm.isActive = true;
    wrapper = shallow(<AlarmItem
      toggleAlarm={toggleAlarmMock}
      alarmId={alarmId}
      navigate={navigate}
      alarm={alarm}
    />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should toggle alarm value', () => {
    alarm.isActive = true;
    wrapper = shallow(<AlarmItem
      toggleAlarm={toggleAlarmMock}
      alarmId={alarmId}
      navigate={navigate}
      alarm={alarm}
    />);
    toggleAlarmMock.mockImplementation(() => new Promise(resolve => resolve()));
    wrapper.find('Switch').simulate(
      'valueChange',
      { preventDefault() {} },
    );
    expect(toggleAlarmMock.mock.calls.length).toBe(1);
  });
});
