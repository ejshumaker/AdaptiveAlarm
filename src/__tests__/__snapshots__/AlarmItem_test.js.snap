// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alarm Item should contain None as no days are active 1`] = `
<TouchableOpacity
  onPress={[Function]}
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginVertical": 0,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "column",
          "justifyContent": "flex-end",
          "marginBottom": 20,
          "marginVertical": 8,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#DDDDDD",
              "fontFamily": "RNSMiles-Bold",
              "fontSize": 48,
            },
            Object {
              "color": "#DDDDDD",
            },
          ]
        }
      >
        8:00 AM
      </Text>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginTop": 5,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#DDDDDD",
                "fontFamily": "RNSMiles-Medium",
                "fontSize": 18,
              },
              Object {
                "color": "#DDDDDD",
                "justifyContent": "flex-start",
              },
            ]
          }
        >
          Repeat:
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#DDDDDD",
                "fontFamily": "RNSMiles-Medium",
                "fontSize": 18,
              },
              Object {
                "color": "#E4F96A",
                "fontFamily": "RNSMiles-XBold",
                "justifyContent": "flex-end",
              },
            ]
          }
        >
           None
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "justifyContent": "space-around",
          "margin": 0,
        }
      }
    >
      <Switch
        onValueChange={[Function]}
        value={true}
      />
    </View>
  </View>
</TouchableOpacity>
`;

exports[`Alarm Item test alarm item matches snapshot alarm is active 1`] = `
<TouchableOpacity
  onPress={[Function]}
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginVertical": 0,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "column",
          "justifyContent": "flex-end",
          "marginBottom": 20,
          "marginVertical": 8,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#DDDDDD",
              "fontFamily": "RNSMiles-Bold",
              "fontSize": 48,
            },
            Object {
              "color": "#DDDDDD",
            },
          ]
        }
      >
        8:00 AM
      </Text>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginTop": 5,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#DDDDDD",
                "fontFamily": "RNSMiles-Medium",
                "fontSize": 18,
              },
              Object {
                "color": "#DDDDDD",
                "justifyContent": "flex-start",
              },
            ]
          }
        >
          Repeat:
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#DDDDDD",
                "fontFamily": "RNSMiles-Medium",
                "fontSize": 18,
              },
              Object {
                "color": "#E4F96A",
                "fontFamily": "RNSMiles-XBold",
                "justifyContent": "flex-end",
              },
            ]
          }
        >
           S, T, W, F,
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "justifyContent": "space-around",
          "margin": 0,
        }
      }
    >
      <Switch
        onValueChange={[Function]}
        value={true}
      />
    </View>
  </View>
</TouchableOpacity>
`;

exports[`Alarm Item test alarm item matches snapshot alarm is inactive 1`] = `
<TouchableOpacity
  onPress={[Function]}
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginVertical": 0,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "column",
          "justifyContent": "flex-end",
          "marginBottom": 20,
          "marginVertical": 8,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#DDDDDD",
              "fontFamily": "RNSMiles-Bold",
              "fontSize": 48,
            },
            Object {
              "color": "#505050",
            },
          ]
        }
      >
        8:00 AM
      </Text>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginTop": 5,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#DDDDDD",
                "fontFamily": "RNSMiles-Medium",
                "fontSize": 18,
              },
              Object {
                "color": "#505050",
                "justifyContent": "flex-start",
              },
            ]
          }
        >
          Repeat:
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#DDDDDD",
                "fontFamily": "RNSMiles-Medium",
                "fontSize": 18,
              },
              Object {
                "color": "#505050",
                "fontFamily": "RNSMiles-XBold",
                "justifyContent": "flex-end",
              },
            ]
          }
        >
           S, T, W, F,
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "justifyContent": "space-around",
          "margin": 0,
        }
      }
    >
      <Switch
        onValueChange={[Function]}
        value={false}
      />
    </View>
  </View>
</TouchableOpacity>
`;
